File: C++/Ternary_Search.cpp
#include<bits/stdc++.h>
using namespace std;

bool ternSearch(vector<int>& arr, int x){
int n=arr.size();
int start=0, end=n-1, mid1,mid2;
while(start<=end){
    mid1=start+(end-start)/3;
    mid2=end-(end-start)/3;
    if(arr[mid1]==x or arr[mid2]==x)
    return true;
    else if(arr[mid1]>x)
    end=mid1-1;
    else if(arr[mid2]<x)
    start=mid2+1;
    else{
        start=mid1+1;
        end=mid2-1;
    }
}
return false;
}
int main()
{
    
    vector<int> arr={12,14, 17, 9, 78, 6};
    int n=arr.size();
    cout<<"Original Array is:"<<endl;
    for(int i=0; i<n ;i++){
        cout<<arr[i]<<" ";
    }
    sort(arr.begin(), arr.end());
    vector<int> search={9, 70, 12, 78};
    for(int i=0; i<search.size(); i++){
    cout<<"Element to be searched in array is: "<<search[i]<<endl;
    if(ternSearch(arr, search[i]))
    cout<<"Element is present";
    else cout<<"Element is absent";
    cout<<endl;
    }

}

File: Java/TernarySearch.java
class TernarySearcher {
 
    static int search(int l, int r, int key, int ar[])
    {
        if (r >= l) {
             int mid1 = l + (r - l) / 3;
            int mid2 = r - (r - l) / 3;
 
             if (ar[mid1] == key) {
                return mid1;
            }
            if (ar[mid2] == key) {
                return mid2;
            }
 
            if (key < ar[mid1]) {
                return search(l, mid1 - 1, key, ar);
            }
            else if (key > ar[mid2]) {
                return search(mid2 + 1, r, key, ar);
            }
            else {
                return search(mid1 + 1, mid2 - 1, key, ar);
            }
        }
        return -1;
    }
    public static void main(String args[])
    {
        int l, r, p, key;
        int ar[] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        l = 0;
        r = 9;
        key = 5;
        p = search(l, r, key, ar);
        System.out.println("Index of " + key + " is " + p);
        key = 50;
        p = search(l, r, key, ar);
        System.out.println("Index of " + key + " is " + p);
    }
}